link/kprobe.go

```diff
@@ -0,0 +1,221 @@
+// Kprobe will also disconnect any Attached eBPF programs.
+func Kprobe(symbol string) (*TraceEvent, error) {
+	return kprobe(symbol, false)
+}
+
+// Kretprobe returns a TraceEvent that fires right before the given kernel
+// symbol exits, with the function stack left intact.
+// See /proc/kallsyms for available symbols. For example, printk():
+//
+//	Kretprobe("printk")
+//
+// As of kernel 4.17, the presence of the given symbol is only checked when
+// Attaching a program later. One TraceEvent can be reused to attach an
+// arbitrary number of programs.
+//
+// Close() must be called on the resulting TraceEvent to clean up system state
+// on kernels before 4.17. On those kernels, Closing a TraceEvent created by
+// Kretprobe will also disconnect any Attached eBPF programs.
+func Kretprobe(symbol string) (*TraceEvent, error) {
+	return kprobe(symbol, true)
+}
+
+// kprobe returns a TraceEvent backed by a k(ret)probe
+// of the given kernel symbol.
+func kprobe(symbol string, ret bool) (*TraceEvent, error) {
+	if symbol == "" {
+		return nil, errors.New("symbol name cannot be empty")
+	}
+	if !rgxTraceEvent.MatchString(symbol) {
+		return nil, fmt.Errorf("symbol must be alphanumeric or underscore: %s", symbol)
+	}
+
+	// Use kprobe PMU if the kernel has it available.
+	tp, err := pmuKprobe(symbol, ret)
+	if err == nil {
+		return tp, nil
+	}
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("creating perf_kprobe PMU: %w", err)
+	}
+
+	// Use tracefs if kprobe PMU is missing.
+	tp, err = tracefsKprobe(symbol, ret)
+	if err != nil {
+		return nil, fmt.Errorf("creating trace event '%s' in tracefs: %w", symbol, err)
+	}
+
+	return tp, nil
+}
+
+// pmuKprobe returns a TraceEvent to be created using PMU (sysfs).
+// Requires at least 4.17 (e12f03d7031a "perf/core: Implement the
+// 'perf_kprobe' PMU").
+func pmuKprobe(symbol string, ret bool) (*TraceEvent, error) {
+	et, err := getPMUEventType("kprobe")
+	if err != nil {
+		return nil, err
+	}
+	return &TraceEvent{
+		pmuID:    et,
+		name:     symbol,
+		ret:      ret,
+		progType: ebpf.Kprobe,
+	}, nil
+}
+
+// tracefsKprobe returns a TraceEvent created by writing an entry to
+// <tracefs>/kprobe_events. On Close(), the entry is removed.
+// A new group identifier is generated on every call to ensure uniqueness.
+func tracefsKprobe(symbol string, ret bool) (*TraceEvent, error) {
+
+	// Generate a random string for each trace event we attempt to create.
+	// This value is used as the 'group' token in tracefs to allow creating
+	// multiple kprobe trace events with the same name.
+	group := mustRandomGroup("ebpf")
+
+	// Before attempting to create a trace event through tracefs,
+	// check if an event with the same group and name already exists.
+	// Kernels 4.x and earlier don't return os.ErrExist on writing a duplicate
+	// entry, so we need to rely on reads for detecting uniqueness.
+	_, err := getTraceEventID(group, symbol)
```

Comment:
This means the code we have to generate a random group isn't random enough? In that case, can we make it more random and drop the check here? If not, should this retry to find a different group?

Reply:
There used to be a retry loop here, but I couldn't justify the complexity I needed to add to be able to test it properly. The value add wasn't clear, because this is never supposed to happen in the first place. Initially, my idea was to allow the caller to specify the group prefix (hardcoded to `ebpf` here), so I kept the random segment to a minimum. Will keep the check here in place, but increase the amount of randomness in the string.

TOCTOU, I know, but this is what we've got to work with on older kernels. Newer kernels will EEXIST when writing to the file anyway, so this is not an issue there. Lowest common denominator.